### PR TITLE
automatic scaling of `epsilon` using `std` instead of `mean` by default.

### PR DESCRIPTION
**Pull request by marcocuturi**
Up to now, `epsilon` was scaled as a fraction of the mean cost matrix, in order to account for scale. The motivation for this was:
- small cost values can handle a smaller `epsilon` value without having `exp(-C/epsilon)` underflow. Likewise, large `C` values also should come with larget `epsilon`.

This intuition is correct when using kernel mode (which requires storing a kernel matrix) but not useful in `logsumexp` mode, which can easily account for costs that are rescaled (e.g. add constant) with no impact computationally speaking.

Indeed, imagine one has an entrywise translation `delta` factor for cost `C`. In that case, the LSE implementation of (balanced) Sinkhorn would handle these two problems in exactly the same way:

`argmin <P, C> - epsilon H(P)` 

and 

`argmin <P, C + delta> - epsilon H(P)= argmin <P, C> + delta - epsilon H(P)` 

(because of coupling `P` total mass = 1).

Yet, the current rule would result in two different `epsilon`'s (specifically the second would be the first `+ 0.05 * delta`, given that we set `epsilon` by default to be equal to 1/20th of the mean cost.

Hence, if "hardness/compute effort" is parameterized by `epsilon`, `epsilon` shouldn't depend on the overall additive scale of `C`, but, instead, on its multiplicative scale. As a result, computing the `std` of the entries in `C` (i.e. the order of magnitude of the centered entries in `C`) is likely to be a more robust alternative.

### CODE REVIEW

**Comment by marcocuturi** [comment]
@michalk8 maybe we could put that `5e-2` factor hardcoded in some "constants.py" file? we could think of other constants, e.g. max/min iterations or default threshold.

File: src/ott/geometry/geometry.py [hunk 1]
```diff
@@ -129,30 +129,58 @@ def mean_cost_matrix(self) -> float:
     tmp = self._masked_geom().apply_cost(self._n_normed_ones).squeeze()
     return jnp.sum(tmp * self._m_normed_ones)
 
+  @property
+  def std_cost_matrix(self) -> float:
+    r"""Standard deviation of all values stored in :attr:`cost_matrix`.
+
+    Uses the :meth:`~ott.geometry.Geometry.apply_square_cost` to remain
```
Comment (michalk8) [inline src/ott/geometry/geometry.py:136]: Should be just :meth:\`apply_square_cost\`; currently it doesn't render correctly.

File: src/ott/geometry/geometry.py [hunk 2]
```diff
@@ -129,30 +129,58 @@ def mean_cost_matrix(self) -> float:
...
+
+    .. math::
+        \sigma^2=\frac{1}{nm}\left(\sum_{ij} C_{ij}^2 -
+        (\sum_{ij}C_ij)^2\right).
+
+    to output :math:`\sigma`.
+    """
+    tmp = self._masked_geom().apply_square_cost(self._n_normed_ones).squeeze()
+    return jnp.sqrt(
+        jax.nn.
```
Comment (michalk8) [inline src/ott/geometry/geometry.py:147]: Can you please split this in a nicer way? The formatting is a bit off.

File: tests/geometry/geometry_test.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+
+from ott.geometry import pointcloud
+
+
+@pytest.mark.fast()
+class TestCostStd:
```
Comment (michalk8) [inline tests/geometry/geometry_test.py:25]: I'd rename this, since you test for both `mean` and `std`.

File: tests/geometry/geometry_test.py
Comment (michalk8) [inline tests/geometry/geometry_test.py:1]: Why not put this in `pointcloud_test.py`?

File: tests/geometry/geometry_test.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+  def test_coststd(self, rng: jax.Array):
```
Comment (michalk8) [inline tests/geometry/geometry_test.py:27]: I'd split this into 2 tests (or parameterize by `stat: Literal["mean", "std"]`.

File: tests/geometry/lr_cost_test.py [hunk 5]
```diff
@@ -284,10 +284,10 @@ def test_apply_transport_from_potentials(self):
     x = jnp.ones((n, d))
     vec, f, g = jnp.ones(n), jnp.zeros(n), jnp.zeros(n)
 
-    geom = low_rank.LRCGeometry(cost_1=x, cost_2=x + 1)
+    geom = low_rank.LRCGeometry(cost_1=x, cost_2=x + 1, epsilon=1.0)
     res = geom.apply_transport_from_potentials(f=f, g=g, vec=vec)
 
-    np.testing.assert_allclose(res, 1.1253539e-7, rtol=1e-6, atol=1e-6)
+    np.testing.assert_allclose(res, 0.183156, rtol=1e-6, atol=1e-6)
```
Comment (michalk8) [inline tests/geometry/lr_cost_test.py:290]: Why was this changed and can we revert?

File: tests/solvers/linear/sinkhorn_test.py [hunk 6]
```diff
@@ -48,6 +48,38 @@ def initialize(self, rng: jax.Array):
     self.a = a / jnp.sum(a)
     self.b = b / jnp.sum(b)
 
+  @pytest.mark.fast.with_args(tau_a=[1.0, 0.93], tau_b=[1.0, 0.91], only_fast=0)
+  def test_lse_matchs(self, tau_a, tau_b):
```
Comment (michalk8) [inline tests/solvers/linear/sinkhorn_test.py:52]: matches

File: tests/solvers/linear/sinkhorn_test.py [hunk 7]
```diff
@@ -48,6 +48,38 @@ def initialize(self, rng: jax.Array):
...
+    """Test that regardless of lse_mode, Sinkhorn returns same value."""
+    geom = pointcloud.PointCloud(self.x, self.y)
+    out = []
+    for lse_mode in [True, False]:
```
Comment (michalk8) [inline tests/solvers/linear/sinkhorn_test.py:56]: Might be nicer to do:
```python
solve_fn = functools.partial(linear.solve, ...)
solve_fn = jax.jit(solve_fn, static_argnames=["lse_mode"])
lse_out = solve_fn(lse_mode=True)
ker_out = solve_fn(lse_mode=False)
```

File: tests/geometry/geometry_test.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    x = jax.random.normal(rngs[0], (n, d))
+    y = jax.random.normal(rngs[1], (m, d)) + 1
+
+    geom = pointcloud.PointCloud(x, y)
+    std = jnp.std(geom.cost_matrix)
+    mean = jnp.mean(geom.cost_matrix)
+    np.testing.assert_allclose(geom.std_cost_matrix, std, rtol=1e-5, atol=1e-5)
+
+    eps = pointcloud.PointCloud(x, y, relative_epsilon="mean").epsilon
+    np.testing.assert_allclose(5e-2 * mean, eps, rtol=1e-5, atol=1e-5)
```
Comment (michalk8) [inline tests/geometry/geometry_test.py:40]: Would extract `5e-2` locally here (for now). Later, we can do as you suggest in constants somewhere in the package.

File: src/ott/geometry/geometry.py [hunk 9]
```diff
@@ -129,30 +129,58 @@ def mean_cost_matrix(self) -> float:
...
+      scale_eps = jax.lax.stop_gradient(self.std_cost_matrix)
+    # If instructions passed change, otherwise (notably if False) skip.
+    elif rel is not None:
+      if rel == "mean" or rel is True:  # Legacy option.
+        scale_eps = jax.lax.stop_gradient(self.mean_cost_matrix)
+      elif rel == "std":
+        scale_eps = jax.lax.stop_gradient(self.std_cost_matrix)
+        # Avoid 0 std, since this would set epsilon to 0.0 and result in
+        # a division by 0.
+        scale_eps = jnp.where(scale_eps <= 0.0, 1.0, scale_eps)
```
Comment (michalk8) [inline src/ott/geometry/geometry.py:177]: Would use `jax.nn.relu(scale_eps)`.

File: src/ott/experimental/mmsinkhorn.py [hunk 10]
```diff
@@ -303,7 +303,7 @@ def __call__(
 
     cost_t = cost_tensor(x_s, cost_fns)
     state = self.init_state(n_s)
-    epsilon = 0.05 * jnp.mean(cost_t) if epsilon is None else epsilon
+    epsilon = 0.05 * jnp.std(cost_t) if epsilon is None else epsilon
```
Comment (michalk8) [inline src/ott/experimental/mmsinkhorn.py:306]: Maybe let's mention it in the docs how the default is set?

File: src/ott/geometry/geometry.py (hunk 2 above)
Comment (marcocuturi) [reply src/ott/geometry/geometry.py:147]: yes, this was yapf. will correct

File: src/ott/geometry/geometry.py (hunk 9 above)
Comment (marcocuturi) [reply src/ott/geometry/geometry.py:177]: that's not the same :)

File: tests/geometry/geometry_test.py
Comment (marcocuturi) [reply tests/geometry/geometry_test.py:1]: i added a test with a regular geometry

File: tests/geometry/geometry_test.py (hunk 4 above)
Comment (marcocuturi) [reply tests/geometry/geometry_test.py:27]: but then there would be lots of "if" no? on both computation of criterion and computation in geom?

File: src/ott/experimental/mmsinkhorn.py (hunk 10 above)
Comment (marcocuturi) [reply src/ott/experimental/mmsinkhorn.py:307]: I have made a pass in various places to ensure this is refreshed.

File: src/ott/geometry/geometry.py (hunk 9 above)
Comment (michalk8) [reply src/ott/geometry/geometry.py:176]: Ah sorry, completely missed it, thank you!

File: src/ott/geometry/epsilon_scheduler.py [hunk 11]
```diff
@@ -18,6 +18,8 @@
 
 __all__ = ["Epsilon"]
```
Comment (michalk8) [inline src/ott/geometry/epsilon_scheduler.py:20]: I think if we add
```python
#: Some description of the constant below.
DEFAULT_SCALE = 0.05
```
and add it to the `.rst` file, we should be able to document it (either using :data: or :obj:).

File: src/ott/experimental/mmsinkhorn.py [hunk 12]
```diff
@@ -303,7 +304,7 @@ def __call__(
...
```
Comment (michalk8) [inline src/ott/experimental/mmsinkhorn.py:307]: Let's also use the `DEFAULT_SCALE` here?

**Review by michalk8 — CHANGES_REQUESTED** [review]

**Comment by michalk8** [comment]
LGTM, thanks a lot @marcocuturi !

**Review by michalk8 — APPROVED** [review]
